internal/networking/state/state.go

```diff
@@ -0,0 +1,175 @@
+		long, lat, err := n.GeolocateIP(ip)
+		if err != nil {
+			log.Printf("Error encountered when looking up coordinates for IP: %v\n\n%v", ip, err)
+			continue // Discard this node
+		}
+		newNode := NewNetworkNode(long, lat, ip)
+		fmt.Printf("Added Node: %v      (%v, %v)\n", ip, long, lat)
+		nodes = append(nodes, newNode)
+	}
+
+	// Create a new KD-Tree with the new set of nodes
+	newTree := kdtree.NewKDTree(nodes)
+	n.mux.Lock()
+	n.tree = newTree
+	n.mux.Unlock()
+}
+
+// GeolocateIP looks up the coordinates for a given ip address
+func (n *NetworkState) GeolocateIP(ip net.IP) (long float64, lat float64, retErr error) {
+	n.mux.Lock()
+	defer n.mux.Unlock()
+	city, err := n.geoIP.City(ip)
+	if err != nil {
+		return 0.0, 0.0, err
+	}
+
+	return city.Location.Longitude, city.Location.Latitude, nil
+}
+
+// GetClosestNode searches the KD-Tree for the node nearest to the IP provided
+func (n *NetworkState) GetClosestNode(ip net.IP) (*NetworkNode, error) {
+	long, lat, err := n.GeolocateIP(ip)
+	if err != nil {
+		return nil, fmt.Errorf("Error encountered when looking up coordinates for IP: %v\n\n%v", ip, err)
+	}
+	n.mux.Lock()
+	defer n.mux.Unlock()
+	// Find the closest neighboring node
+	neighbors := n.tree.KNN(NewNetworkNode(long, lat, ip), 1)
+	if len(neighbors) == 0 {
+		return nil, fmt.Errorf("Error: No neighbors were found in nearest neighbor search")
+	}
+	return neighbors[0].(*NetworkNode), nil
+}
+
+// RunningStateChanged returns a channel that updates when the running state
+// is changed
+func (n *NetworkState) RunningStateChanged() chan (bool) {
+	return n.runChannel
+}
+
+// NetworkNode is a struct representing an edge node in a pool
+type NetworkNode struct {
+	kdtree.Point // Implements Point interface
+
+	geoLocation s2.LatLng
+	ip          net.IP
+}
+
+// NewNetworkNode returns a new NetworkNode struct
+func NewNetworkNode(lat, long float64, ip net.IP) *NetworkNode {
+	return &NetworkNode{geoLocation: s2.LatLngFromDegrees(lat, long), ip: ip}
+}
+
+func (n *NetworkNode) String() string {
+	return fmt.Sprintf("%v [%v, %v]", n.ip, n.geoLocation.Lat.Degrees(), n.geoLocation.Lng.Degrees())
+}
+
+// IP returns the IP address of the NetworkNode
+func (n *NetworkNode) IP() net.IP {
+	return n.ip
+}
+
+// Dim returns the number of dimensions the KD-Tree splits on
+func (n *NetworkNode) Dim() int {
+	return 2
+}
+
+// GetValue returns the value associated with the provided dimension
+func (n *NetworkNode) GetValue(dim int) float64 {
+	if dim == 0 {
+		return n.geoLocation.Lat.Degrees()
+	}
+	return n.geoLocation.Lng.Degrees()
+}
+
+// Distance returns the calculated distance between two nodes
+func (n *NetworkNode) Distance(other kdtree.Point) float64 {
+	return n.geoLocation.Distance(other.(*NetworkNode).geoLocation).Radians()
+}
+
+// PlaneDistance returns the distance between the point and the specified
+// plane
+func (n *NetworkNode) PlaneDistance(val float64, dim int) float64 {
```

Comment:
Should probably do a different distance function to deal with earth being round not a plane